### PR TITLE
SERVER-88082 Add workloads for WriteType::WriteWithoutShardKeyWithId

### PR DESCRIPTION
**Pull request by zorro786**
**Jira Ticket:** SERVER-88082

**Whats Changed:**  
Added two new workloads to measure performance under PM-3190

**Patch testing results:**  
(https://spruce.mongodb.com/version/65f4b6c4c50c4400078120c8/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC)

**Workload Submission form:**  
[Pre-filled form](https://docs.google.com/forms/d/e/1FAIpQLSf1oh23Ddo1khjLZMqZ9DHbRdObnpeH20PSXTBuXVg-7mxxTA/viewform?usp=pp_url&entry.1544049958=PM-3190.+This+workload+tests+the+performance+of+batched+writes+without+shard+key+with+_id.&entry.621283182=Master+Only&entry.132161754=sys-perf&entry.1936323232=Linux+Shard+Lite+Cluster+2022-11(2+shards,+2+node+replica+sets,+once+per+day)&entry.1936323232=Linux+3-Shard+Cluster+2022-11+(once+per+week)&entry.1936323232=Linux+Shard+Lite+All+Feature+Flags+2022-11&entry.1482827051=genny+(https://github.com/mongodb/genny)&entry.349286116=UpdateOneMetrics+throughput+and+latency+(P50,+P99)&entry.361971069=This+is+a+new+workload+so+we+would+like+to+establish+a+base+line.+We+have+however+tested+pre+PM-3190+and+post+PM-3190+workloads+and+found+the+comparison+to+be+within+scope+https://docs.google.com/spreadsheets/d/17Tq7zUsB2y86t9gjH2a2WrGuqIZH9KkTuqCnR1lpBp4/edit?userstoinvite%3Dgaraudy.etienne@mongodb.com%26sharingaction%3Dmanageaccess%26role%3Dwriter%23gid%3D0&entry.1838923236=UpdateOneMetrics+throughput+and+latency+(P50,+P99)&entry.605497846=No&entry.538491135=N/A)

### CODE REVIEW

File: src/workloads/sharding/BatchedUpdateOneWithoutShardKeyWithId.yml [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    1. Creating an empty sharded collection distributed across all shards in the cluster.
+    2. Populating the sharded collection with data.
+    3. Running updateOne operations of type WriteWithoutShardKeyWithId.
+
+  The inserted documents have the following form:
+
+      {_id: 10, oldKey: 20, newKey: 30, counter: 0, padding: 'random string of bytes ...'}
+
+  The collection is sharded on {oldKey: 'hashed'}.
+
```
Comment (jimoleary) [inline src/workloads/sharding/BatchedUpdateOneWithoutShardKeyWithId.yml:16]: Can you add a Keywords section?

File: src/workloads/sharding/BatchedUpdateOneWithoutShardKeyWithId.yml [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+  The workload consists of 3 phases:
+    1. Creating an empty sharded collection distributed across all shards in the cluster.
+    2. Populating the sharded collection with data.
+    3. Running updateOne operations of type WriteWithoutShardKeyWithId.
+
+  The inserted documents have the following form:
+
+      {_id: 10, oldKey: 20, newKey: 30, counter: 0, padding: 'random string of bytes ...'}
+
+  The collection is sharded on {oldKey: 'hashed'}.
```
Comment (jimoleary) [inline src/workloads/sharding/BatchedUpdateOneWithoutShardKeyWithId.yml:15]: What are the important metrics for this workload?

File: src/workloads/sharding/BatchedUpdateOneWithoutShardKeyWithId.yml [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+  Runs the batched updateOne writes of type WithoutShardKeyWithId with a batch size of 1000.
+
+  The workload consists of 3 phases:
+    1. Creating an empty sharded collection distributed across all shards in the cluster.
+    2. Populating the sharded collection with data.
+    3. Running updateOne operations of type WriteWithoutShardKeyWithId.
+
+  The inserted documents have the following form:
+
+      {_id: 10, oldKey: 20, newKey: 30, counter: 0, padding: 'random string of bytes ...'}
```
Comment (jimoleary) [inline src/workloads/sharding/BatchedUpdateOneWithoutShardKeyWithId.yml:13]: nit: ```padding: 'random string of bytes to bring the docs up to 8 to 10 KB...'```

File: src/workloads/sharding/BatchedUpdateOneWithoutShardKeyWithId.yml [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+  - &ShardKeyValueMin 1
+  - &ShardKeyValueMax 1000
+
+  - &Document {
+      oldKey: {^RandomInt: {min: *ShardKeyValueMin, max: *ShardKeyValueMax}},
+      newKey: {^RandomInt: {min: *ShardKeyValueMin, max: *ShardKeyValueMax}},
+      counter: 0,
+    }
+
+  - &WriteOp {
```
Comment (jimoleary) [inline src/workloads/sharding/BatchedUpdateOneWithoutShardKeyWithId.yml:40]: nit: UpdateOps?

File: src/workloads/sharding/BatchedUpdateOneWithoutShardKeyWithId.yml [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+  Type: MonotonicSingleLoader
+  Threads: 100
+  Phases:
+  - *Nop
+  - Repeat: 1
+    BatchSize: 1000
+    DocumentCount: *DocumentCount
+    Database: *Database
+    Document:
+      oldKey: {^RandomInt: {min: *ShardKeyValueMin, max: *ShardKeyValueMax}}
```
Comment (jimoleary) [inline src/workloads/sharding/BatchedUpdateOneWithoutShardKeyWithId.yml:102]: Suggestion: These values (```{^RandomInt: {min: *ShardKeyValueMin, max: *ShardKeyValueMax}}```) are also used in [&Document](https://github.com/mongodb/genny/pull/1151/files#diff-d51969533f015d861134ae4d73baa44f94b0cb6f4f5f34a79abbd5bffeca16aaR34-R38). It probably makes sense to consolidate the defintions in one place and reuse them here.

File: src/workloads/sharding/BulkWriteBatchedUpdateOneWithoutShardKeyWithId.yml [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+    1. Creating an empty sharded collection distributed across all shards in the cluster.
+    2. Populating the sharded collection with data.
+    3. Running updateOne writes of type WriteWithoutShardKeyWithId.
+
+  The inserted documents have the following form:
+
+      {_id: 10, oldKey: 20, newKey: 30, counter: 0, padding: 'random string of bytes ...'}
+
+  The collection is sharded on {oldKey: 'hashed'}.
+
```
Comment (jimoleary) [inline src/workloads/sharding/BulkWriteBatchedUpdateOneWithoutShardKeyWithId.yml:17]: The rest of the comments from BatchedUpdateOneWithoutShardKeyWithId.yml also apply here.

File: src/workloads/sharding/BulkWriteBatchedUpdateOneWithoutShardKeyWithId.yml [hunk 7]
```diff
@@ -0,0 +1,137 @@
+SchemaVersion: 2018-07-01
+Owner: "@mongodb/sharding"
+Description: |
+  Runs the batched updateOne writes of type WithoutShardKeyWithId with a batch size of 1000 using
```
Comment (jimoleary) [inline src/workloads/sharding/BulkWriteBatchedUpdateOneWithoutShardKeyWithId.yml:4]: 'Runs the **bulk** updateOne writes ...'

File: src/workloads/sharding/BatchedUpdateOneWithoutShardKeyWithId.yml [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+  # Note that the exact document size may exceed ApproxDocumentSize because of field names and other
+  # fields in the document.
+  - &ApproxDocumentSize 10000  # = 10kB
+  - &ApproxDocumentSize80Pct 8000  # = 8kB
+  - &DocumentCount 100000  # for an approximate total of 1GB
+
+  - &ShardKeyValueMin 1
+  - &ShardKeyValueMax 1000
+
+  - &Document {
```
Comment (jimoleary) [inline src/workloads/sharding/BatchedUpdateOneWithoutShardKeyWithId.yml:34]: The Document anchor isn't used anywhere and is incomplete? Maybe fill out the rest of it and use it in the loader or remove it (you already explain the structure in the comment at the top).

**Review by jimoleary — COMMENTED** [review]
Overall LGTM, I've left some comments for some more documentation. Generally the comments apply to both files.

Question: Is it worth extracting most of the common yaml into a file in workloads/phases somewhere? The workloads are small so I'm ambivalent.

File: src/workloads/sharding/BatchedUpdateOneWithoutShardKeyWithId.yml (hunk 1 above)
Comment (zorro786) [reply src/workloads/sharding/BatchedUpdateOneWithoutShardKeyWithId.yml:26]: Done.

File: src/workloads/sharding/BatchedUpdateOneWithoutShardKeyWithId.yml (hunk 2 above)
Comment (zorro786) [reply src/workloads/sharding/BatchedUpdateOneWithoutShardKeyWithId.yml]: Added, let me know if it suffices or if I need to write their names.

File: src/workloads/sharding/BatchedUpdateOneWithoutShardKeyWithId.yml (hunk 3 above)
Comment (zorro786) [reply src/workloads/sharding/BatchedUpdateOneWithoutShardKeyWithId.yml]: Done.

File: src/workloads/sharding/BatchedUpdateOneWithoutShardKeyWithId.yml (hunk 4 above)
Comment (zorro786) [reply src/workloads/sharding/BatchedUpdateOneWithoutShardKeyWithId.yml]: Done.

File: src/workloads/sharding/BulkWriteBatchedUpdateOneWithoutShardKeyWithId.yml (hunk 6 above)
Comment (zorro786) [reply src/workloads/sharding/BulkWriteBatchedUpdateOneWithoutShardKeyWithId.yml:28]: Done.

File: src/workloads/sharding/BatchedUpdateOneWithoutShardKeyWithId.yml (hunk 8 above)
Comment (zorro786) [reply src/workloads/sharding/BatchedUpdateOneWithoutShardKeyWithId.yml]: Removed.

File: src/workloads/sharding/BulkWriteBatchedUpdateOneWithoutShardKeyWithId.yml (hunk 7 above)
Comment (zorro786) [reply src/workloads/sharding/BulkWriteBatchedUpdateOneWithoutShardKeyWithId.yml:4]: Hmm so it actually is batched in this case too but done differently (using bulkWrite command)

**Comment by zorro786** [comment]
>  Is it worth extracting most of the common yaml into a file in workloads/phases somewhere? The workloads are small so I'm ambivalent.

I would prefer to keep them verbose within a file as they are not that big - when debugging or understanding workloads referring to multiple files makes it a bit confusing/tedious.

File: src/workloads/sharding/BulkWriteBatchedUpdateOneWithoutShardKeyWithId.yml (hunk 7 above)
Comment (jimoleary) [reply src/workloads/sharding/BulkWriteBatchedUpdateOneWithoutShardKeyWithId.yml:4]: Maybe 'Runs the batched bulk updateOne writes ...' then ?

**Review by jimoleary — APPROVED** [review]
> I would prefer to keep them verbose within a file as they are not that big - when debugging or understanding workloads referring to multiple files makes it a bit confusing/tedious.

No problem, that seems reasonable to me given the size / scale of these workloads.

LGTM ... there are 2 very-nit-picky comments left ... feel free to address them or not.

File: src/workloads/sharding/BatchedUpdateOneWithoutShardKeyWithId.yml (hunk 5 above)
Comment (zorro786) [reply src/workloads/sharding/BatchedUpdateOneWithoutShardKeyWithId.yml:106]: I removed `&Document` as suggested in another comment.

File: src/workloads/sharding/BulkWriteBatchedUpdateOneWithoutShardKeyWithId.yml (hunk 7 above)
Comment (zorro786) [reply src/workloads/sharding/BulkWriteBatchedUpdateOneWithoutShardKeyWithId.yml]: Done.

**Comment by zorro786** [comment]
evergreen merge

**Pull request merged**